File: src/flutter/flet_animated_border/lib/src/flet_animated_border.dart
import 'package:flet/flet.dart';
import 'package:flutter/material.dart';
import 'package:zo_animated_border/zo_animated_border.dart';

class FletAnimatedBorderControl extends StatefulWidget {
  final Control? parent;
  final Control control;
  final bool parentDisabled;
  final bool? parentAdaptive;
  final List<Control> children;
  final FletControlBackend backend;

  const FletAnimatedBorderControl({
    super.key,
    this.parent,
    required this.control,
    required this.children,
    required this.parentDisabled,
    required this.parentAdaptive,
    required this.backend,
  });
  @override
  State<FletAnimatedBorderControl> createState() =>
      _FletAnimatedBorderControlState();
}

class _FletAnimatedBorderControlState extends State<FletAnimatedBorderControl>
    with TickerProviderStateMixin {
  double _prevBorderWidth = 8.0;
  double _prevBorderRadius = 10.0;
  bool _isFirstBuild = true;

  late AnimationController _controller;
  late Animation<double> _borderWidthAnim;
  late Animation<double> _borderRadiusAnim;

  @override
  void initState() {
    super.initState();

    _controller = AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 300),
    );
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  void _setupAndStartAnimations(
      double targetWidth, double targetRadius, Duration duration, Curve curve) {
    _borderWidthAnim = Tween<double>(
      begin: _prevBorderWidth,
      end: targetWidth,
    ).animate(CurvedAnimation(parent: _controller, curve: curve));

    _borderRadiusAnim = Tween<double>(
      begin: _prevBorderRadius,
      end: targetRadius,
    ).animate(CurvedAnimation(parent: _controller, curve: curve));

    _prevBorderWidth = targetWidth;
    _prevBorderRadius = targetRadius;

    _controller.reset();
    _controller.duration = duration;
    _controller.forward();
  }

  @override
  Widget build(BuildContext context) {
    debugPrint("AnimatedBorder build: ${widget.control.id}");

    bool disabled = widget.control.isDisabled || widget.parentDisabled;

    final targetBorderWidth = widget.control.attrDouble("borderWidth") ?? 8.0;
    final targetBorderRadius =
        widget.control.attrDouble("borderRadius") ?? 10.0;
    final glowOpacity = widget.control.attrDouble("glowOpacity") ?? 0.4;
    final durationSeconds = widget.control.attrInt("duration_seconds") ?? 3;
    final borderType = widget.control.attrString("borderType") ?? "dual";
    final smoothGradientLoop =
        widget.control.attrBool("smoothGradientLoop", true)!;

    // Extract gradient colors directly from the 'gradientColors' attribute
    List<Color> gradientColors = [];

    // Try to get gradient colors list directly
    final gradientColorsJson = widget.control.attrList("gradientColors");
    if (gradientColorsJson != null) {
      gradientColors = gradientColorsJson
          .map((colorStr) =>
              parseColor(Theme.of(context), colorStr.toString()) ??
              Colors.white)
          .toList();
    }

    // If no gradient colors provided, use dual colors as fallback

    var animation = parseAnimation(widget.control, "animate");
    final animDuration =
        animation?.duration ?? const Duration(milliseconds: 300);
    final animCurve = animation?.curve ?? Curves.easeInOut;

    debugPrint("AnimatedBorder params: borderRadius=$targetBorderRadius, "
        "borderWidth=$targetBorderWidth, glowOpacity=$glowOpacity, "
        "borderType=$borderType, gradientColors=${gradientColors.length} colors");

    if (animation != null) {
      debugPrint(
          "Animation: duration=${animDuration.inMilliseconds}ms, curve=$animCurve");
    }

    bool hasChanges = !_isFirstBuild &&
        (_prevBorderWidth != targetBorderWidth ||
            _prevBorderRadius != targetBorderRadius ||
            // Force rebuild when colors change to ensure they update properly
            true);

    if (_isFirstBuild) {
      _prevBorderWidth = targetBorderWidth;
      _prevBorderRadius = targetBorderRadius;
      _isFirstBuild = false;

      _borderWidthAnim = AlwaysStoppedAnimation<double>(targetBorderWidth);
      _borderRadiusAnim = AlwaysStoppedAnimation<double>(targetBorderRadius);
    } else if (hasChanges && animation != null) {
      debugPrint("Starting animation for borderWidth and borderRadius");

      WidgetsBinding.instance.addPostFrameCallback((_) {
        _setupAndStartAnimations(
            targetBorderWidth, targetBorderRadius, animDuration, animCurve);
      });
    } else if (hasChanges) {
      _prevBorderWidth = targetBorderWidth;
      _prevBorderRadius = targetBorderRadius;

      _borderWidthAnim = AlwaysStoppedAnimation<double>(targetBorderWidth);
      _borderRadiusAnim = AlwaysStoppedAnimation<double>(targetBorderRadius);
    }

    var contentCtrls =
        widget.children.where((c) => c.name == "content" && c.isVisible);

    Widget? child = contentCtrls.isNotEmpty
        ? createControl(widget.control, contentCtrls.first.id, disabled,
            parentAdaptive: false)
        : null;

    final onAnimationEnd = widget.control.attrBool("onAnimationEnd", false)!
        ? () {
            widget.backend.triggerControlEvent(
                widget.control.id, "animation_end", "animated_border");
          }
        : null;

    if (onAnimationEnd != null) {
      _controller.removeStatusListener(_handleAnimationStatus);
      _controller.addStatusListener(_handleAnimationStatus);
    }

    return AnimatedBuilder(
      animation: _controller,
      builder: (context, _) {
        final currentWidth = _borderWidthAnim.value;
        final currentRadius = _borderRadiusAnim.value;

        final borderRadiusObj = BorderRadius.circular(currentRadius);

        final bool showBorder = currentWidth > 0;

        if (!showBorder) {
          Widget contentWidget = Container(
            alignment: Alignment.center,
            decoration: BoxDecoration(
              color: Colors.transparent,
              borderRadius: borderRadiusObj,
            ),
            child: child ??
                const Text("Animated Border",
                    style: TextStyle(color: Colors.black)),
          );

          return constrainedControl(
              context, contentWidget, widget.parent, widget.control);
        }

        Widget baseWidget;

        // Choose border type based on borderType parameter
        if (borderType == "soft_gradient") {
          // Use ZoAnimatedGradientBorder
          var smoothGradientColors = [...gradientColors];

          // Fix for smooth gradient loop transition
          if (smoothGradientLoop && smoothGradientColors.length > 1) {
            // For a smooth loop, add the first color at the end
            // regardless of how many colors are in the list
            smoothGradientColors.add(smoothGradientColors[0]);
            debugPrint("Added first color to end for smooth gradient loop");
          }

          baseWidget = ZoAnimatedGradientBorder(
            borderRadius: currentRadius,
            borderThickness: currentWidth,
            gradientColor: smoothGradientColors,
            glowOpacity: glowOpacity,
            duration: Duration(seconds: durationSeconds),
            child: Container(
              alignment: Alignment.center,
              decoration: BoxDecoration(
                color: Colors.transparent,
                borderRadius: borderRadiusObj,
              ),
              child: child ??
                  const Text("Gradient Border",
                      style: TextStyle(color: Colors.black)),
            ),
          );
        } else {
          // Default to ZoDualBorder
          baseWidget = ZoDualBorder(
            duration: Duration(seconds: durationSeconds),
            glowOpacity: glowOpacity,
            firstBorderColor:
                widget.control.attrColor("firstDualColor", context) ??
                    Colors.blue,
            secondBorderColor:
                widget.control.attrColor("secondDualColor", context) ??
                    Colors.red,
            trackBorderColor:
                widget.control.attrColor("trackDualColor", context) ??
                    Colors.grey,
            borderWidth: currentWidth,
            borderRadius: borderRadiusObj,
            child: Container(
              alignment: Alignment.center,
              decoration: BoxDecoration(
                color: Colors.transparent,
                borderRadius: borderRadiusObj,
              ),
              child: child ??
                  const Text("Dual Border",
                      style: TextStyle(color: Colors.black)),
            ),
          );
        }

        return constrainedControl(
            context, baseWidget, widget.parent, widget.control);
      },
    );
  }

  void _handleAnimationStatus(AnimationStatus status) {
    if (status == AnimationStatus.completed) {
      widget.backend.triggerControlEvent(
          widget.control.id, "animation_end", "animated_border");
    }
  }
}

File: src/flet_animated_border/flet_animated_border.py
from enum import Enum
from typing import Any, Optional, List, Union

from flet.core.constrained_control import ConstrainedControl
from flet.core.control import Control, OptionalNumber
from flet.core.animation import AnimationValue
from flet.core.gradients import Gradient

from flet.core.types import (
    ColorEnums,
    ColorValue,
)
class BorderType(Enum):
    DUAL = "dual"
    SOFT_GRADIENT = "soft_gradient"


class AnimatedBorder(ConstrainedControl):
    """
    AnimatedBorder Control description.
    """

    def __init__(
        self,
        #
        # Control
        #
        content: Optional[Control] = None,
        opacity: OptionalNumber = None,
        tooltip: Optional[str] = None,
        visible: Optional[bool] = None,
        data: Any = None,
        #
        # ConstrainedControl
        #
        left: OptionalNumber = None,
        top: OptionalNumber = None,
        right: OptionalNumber = None,
        bottom: OptionalNumber = None,
        #
        # FletAnimatedBorder specific
        #
        borderWidth: Optional[float] = 5.0,
        borderRadius: Optional[float] = 5.0,
        glowOpacity: Optional[float] = 0.0,
        firstDualColor: Optional[str] = "yellow",
        secondDualColor: Optional[str] = "orange",
        trackDualColor: Optional[str] = "transparent",
        border_type: Optional[BorderType] = BorderType.DUAL,
        gradient_colors: Optional[List[str]] = None,
        smooth_gradient_loop: Optional[bool] = True,
        duration_seconds: Optional[int] = 3,
        animate: Optional[AnimationValue] = None,
        on_animation_end: Optional[callable] = None,
    ):
        ConstrainedControl.__init__(
            self,
            tooltip=tooltip,
            opacity=opacity,
            visible=visible,
            data=data,
            left=left,
            top=top,
            right=right,
            bottom=bottom,
        )

        self._border_width = None
        self._border_radius = None
        self._glow_opacity = None
        self._duration_seconds = None
        self._animate = None
        self._on_animation_end = None
        self._first_dual_color = None
        self._second_dual_color = None
        self._track_dual_color = None
        self._border_type = None
        self._gradient = None
        self._gradient_colors = None
        self._smooth_gradient_loop = None
        
        self.__content = None
        
        self.content = content
        self.borderWidth = borderWidth
        self.borderRadius = borderRadius
        self.glowOpacity = glowOpacity
        self.firstDualColor = firstDualColor
        self.secondDualColor = secondDualColor
        self.trackDualColor = trackDualColor
        self.border_type = border_type
        self.gradient_colors = gradient_colors
        self.smooth_gradient_loop = smooth_gradient_loop
        self.duration_seconds = duration_seconds
        self.animate = animate
        self.on_animation_end = on_animation_end

    def _get_control_name(self):
        return "flet_animated_border"

    def before_update(self):
        super().before_update()
        self._set_attr_json("borderWidth", self._border_width)
        self._set_attr_json("borderRadius", self._border_radius)
        self._set_attr_json("glowOpacity", self._glow_opacity)
        self._set_attr_json("duration_seconds", self._duration_seconds)
        self._set_attr_json("animate", self._animate)
        self._set_attr("onAnimationEnd", True if self._on_animation_end is not None else None)
        self._set_attr("borderType", self._border_type.value if self._border_type else None)
        self._set_attr("smoothGradientLoop", self._smooth_gradient_loop)
        
        # Extract colors for ZoAnimatedGradientBorder
        gradient_colors_list = self._gradient_colors or []
        
        # If a gradient is provided but no explicit gradient_colors, extract colors from gradient
        if not gradient_colors_list and self._gradient and hasattr(self._gradient, "colors"):
            gradient_colors_list = self._gradient.colors
            
        # Send gradient colors to Dart implementation
        self._set_attr_json("gradientColors", gradient_colors_list)
    
    def _get_children(self):
        children = []
        if self.__content is not None:
            self.__content._set_attr_internal("n", "content")
            children.append(self.__content)
        return children

    # content
    @property
    def content(self) -> Optional[Control]:
        return self.__content

    @content.setter
    def content(self, value: Optional[Control]):
        self.__content = value

    # borderWidth
    @property
    def borderWidth(self):
        return self._border_width
    
    @borderWidth.setter
    def borderWidth(self, value):
        print(f"Setting borderWidth to {value}")
        self._border_width = value
    
    # borderRadius
    @property
    def borderRadius(self):
        return self._border_radius
    
    @borderRadius.setter
    def borderRadius(self, value):
        print(f"Setting borderRadius to {value}")
        self._border_radius = value
    
    # glowOpacity
    @property
    def glowOpacity(self):
        return self._glow_opacity

    @glowOpacity.setter
    def glowOpacity(self, value):
        print(f"Setting glowOpacity to {value}")
        self._glow_opacity = value
        
    # duration_seconds
    @property
    def duration_seconds(self):
        return self._duration_seconds

    @duration_seconds.setter
    def duration_seconds(self, value):
        print(f"Setting duration_seconds to {value}")
        self._duration_seconds = value
        
    # animate
    @property
    def animate(self) -> Optional[AnimationValue]:
        return self._animate

    @animate.setter
    def animate(self, value: Optional[AnimationValue]):
        print(f"Setting animate to {value}")
        self._animate = value
        
    # on_animation_end
    @property
    def on_animation_end(self):
        return self._on_animation_end

    @on_animation_end.setter
    def on_animation_end(self, handler):
        self._on_animation_end = handler
        self._add_event_handler("animation_end", handler)
        
    # firstDualColor
    @property
    def firstDualColor(self):
        return self._first_dual_color
    
    @firstDualColor.setter
    def firstDualColor(self, value: Optional[ColorValue]):
        self._first_dual_color = value
        self._set_enum_attr("firstDualColor", value, ColorEnums)
        
    # secondDualColor
    @property
    def secondDualColor(self):
        return self._second_dual_color
    
    @secondDualColor.setter
    def secondDualColor(self, value: Optional[ColorValue]):
        self._second_dual_color = value
        self._set_enum_attr("secondDualColor", value, ColorEnums)
        
    # trackDualColor
    @property
    def trackDualColor(self):
        return self._track_dual_color
    
    @trackDualColor.setter
    def trackDualColor(self, value: Optional[ColorValue]):
        self._track_dual_color = value
        self._set_enum_attr("trackDualColor", value, ColorEnums)
        
    # border_type
    @property
    def border_type(self) -> Optional[BorderType]:
        return self._border_type
    
    @border_type.setter
    def border_type(self, value: Optional[BorderType]):
        print(f"Setting border_type to {value}")
        self._border_type = value
        
    # gradient_colors
    @property
    def gradient_colors(self) -> Optional[List[str]]:
        return self._gradient_colors
    
    @gradient_colors.setter
    def gradient_colors(self, value: Optional[List[str]]):
        print(f"Setting gradient_colors to {value}")
        self._gradient_colors = value
        
    # smooth_gradient_loop
    @property
    def smooth_gradient_loop(self) -> Optional[bool]:
        return self._smooth_gradient_loop
    
    @smooth_gradient_loop.setter
    def smooth_gradient_loop(self, value: Optional[bool]):
        print(f"Setting smooth_gradient_loop to {value}")
        self._smooth_gradient_loop = value
